polkadot/node/core/pvf/common/src/worker/security/clone.rs

```diff
@@ -0,0 +1,92 @@
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Functionality for securing the job processes spawned by the workers using `clone`. If
+//! unsupported, falls back to `fork`.
+
+use crate::{worker::WorkerInfo, LOG_TARGET};
+use nix::{
+	errno::Errno,
+	sched::{CloneCb, CloneFlags},
+	unistd::Pid,
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+	#[error("could not clone, errno: {0}")]
+	Clone(Errno),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Try to run clone(2) on the current worker.
+///
+/// SAFETY: new process should be either spawned within a single threaded process, or use only
+/// async-signal-safe functions.
+pub unsafe fn clone_on_worker(
+	worker_info: &WorkerInfo,
+	cb: CloneCb,
+	stack_size: usize,
+) -> Result<Pid> {
+	let flags = clone_flags();
+
+	gum::trace!(
+		target: LOG_TARGET,
+		?worker_info,
+		%stack_size,
+		"calling clone with flags: {:?}",
+		flags
+	);
+
+	try_clone(cb, stack_size, flags)
+}
+
+/// Runs a check for clone(2) with all sandboxing flags and returns an error indicating whether it
+/// can be fully enabled on the current Linux environment.
+///
+/// SAFETY: new process should be either spawned within a single threaded process, or use only
+/// async-signal-safe functions.
+pub unsafe fn check_can_fully_clone() -> Result<()> {
+	let stack_size = 2 * 1024 * 1024; // Use same as prepare worker for this check.
+	try_clone(Box::new(|| 0), stack_size, clone_flags()).map(|_pid| ())
+}
+
+/// Runs clone(2) with all sandboxing flags.
+///
+/// SAFETY: new process should be either spawned within a single threaded process, or use only
+/// async-signal-safe functions.
+unsafe fn try_clone(cb: CloneCb, stack_size: usize, flags: CloneFlags) -> Result<Pid> {
+	let mut stack: Vec<u8> = vec![0u8; stack_size];
+
+	nix::sched::clone(cb, stack.as_mut_slice(), flags, None).map_err(|errno| Error::Clone(errno))
+}
+
+/// Returns flags for `clone(2)`, including all the sandbox-related ones.
+fn clone_flags() -> CloneFlags {
+	// NOTE: CLONE_NEWUSER does not work when cloning job processes, but in Secure Validator Mode it
+	// is already set by the worker.
+	//
+	// SIGCHLD flag is used to inform clone that the parent process is
+	// expecting a child termination signal, without this flag `waitpid` function
+	// return `ECHILD` error.
+	CloneFlags::CLONE_NEWCGROUP |
```

Comment:
```
From man page:
Only a privileged process (CAP_SYS_ADMIN) can employ CLONE_NEWCGROUP
```
Is that something we want ?
 
Not sure if that is already a hard requirement for running `polkadot` but it feels like it shouldn't be.

Reply:
Good catch! I was going to include the `CLONE_NEWUSER` clone flag, which gives us this capability. From `unshare(2)`:

              As with the child process created by [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html) with the
              CLONE_NEWUSER flag, the caller obtains a full set of
              capabilities in the new namespace.

However, for reasons I don't understand, `clone` with this flag fails if we already passed the same flag to `unshare` [here](https://github.com/paritytech/polkadot-sdk/blob/69713a718566a753ba2735cd9810546f6c0eb681/polkadot/node/core/pvf/common/src/worker/security/change_root.rs#L94), when sandboxing the parent worker process. So I was going to parametrize this function over whether the `unshare` capability is present - I just forgot to do it. If it is present, then the parent worker process already has `CAP_SYS_ADMIN`. That's also why`clone_flags` is a `fn` and not a `const` @alindima.

Reply:
Okay, made this change. If for some reason we cannot `unshare`, we will need the `CAP_SYS_ADMIN` so I added the clone flag for that case. Hopefully there are no issues doing so in the wild, because admittedly local testing and CI have not caught all the issues in the past. However, this new capability is optional, so worst-case scenario someone sees a warning and reports it to us.